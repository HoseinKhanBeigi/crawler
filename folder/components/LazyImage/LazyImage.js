/* eslint-disable jsx-a11y/click-events-have-key-events,jsx-a11y/interactive-supports-focus,jsx-a11y/no-noninteractive-element-interactions */
import React from 'react';
import PropTypes from 'prop-types';
import cs from 'classnames';
import withStyles from 'isomorphic-style-loader/lib/withStyles';
import { connect } from 'react-redux';
import saveAs from 'file-saver';
import { Icon } from 'antd';
import s from './LazyImage.scss';
import {
  getDocAction,
  getFileTokenAction,
} from '../../store/docDownload/docDownload.actions';
import ImageTools from '../ImageTools';
import CPTooltip from '../CP/CPTooltip';

class LazyImage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      loaded: false,
      error: false,
      objectURL: null,
      pdfObjectUrl: null,
    };
  }

  componentDidMount() {
    this.getImage();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.path !== this.props.path) {
      this.getImage();
    }
  }

  getImage = async () => {
    try {
      const { getTokenBefore, path, objectToken } = this.props;
      let token = null;
      if (getTokenBefore) {
        const pathBodyObj = {
          names: [path],
        };
        token = await this.props.getFileTokenAction(pathBodyObj);
      }

      if ((objectToken || getTokenBefore) && path) {
        const blob = await this.props.getDocAction({
          path,
          objectToken: token?.resp?.token || objectToken?.token,
        });
        const objectURL = await URL.createObjectURL(blob);
        this.setState({ pdfObjectUrl: blob });
        if (objectURL) {
          const img = new Image();
          img.onload = () => {
            this.setState({
              loaded: true,
              objectURL,
            });
          };
          img.onerror = () => {
            this.setState({
              error: true,
            });
          };
          img.src = objectURL;
        }
      }
    } catch (e) {
      // eslint-disable-next-line no-console
      console.log(e);
    }
  };

  handelDownload = () => {
    const { objectURL } = this.state;
    const { path } = this.props;
    const slashIndex = path.lastIndexOf('/');
    const originalName = path.slice(slashIndex);

    if (objectURL) {
      saveAs(objectURL, `${originalName}`);
    }
  };

  sendImagePath = () => {};

  handelDownloadPdf = () => {
    const { pdfObjectUrl } = this.state;
    const { path } = this.props;
    const slashIndex = path.lastIndexOf('/');
    const originalName = path.slice(slashIndex);
    saveAs(pdfObjectUrl, originalName);
  };
  renderImg = () => {
    const { download, ref, className, hasZoom, path } = this.props;
    const { error, loaded, objectURL } = this.state;
    if (error || !path) {
      return (
        <div className={s.noData}>
          <CPTooltip title="دانلود این فایل">
            <Icon
              style={{ padding: '10px', fontSize: '18px', color: '#1890ff' }}
              type="download"
              onClick={this.handelDownloadPdf}
            />
          </CPTooltip>
          {/* <Icon type="warning" /> */}
        </div>
      );
    } else if (!loaded) {
      return (
        <div className={s.noData}>
          <Icon type="loading" />
        </div>
      );
    }

    return hasZoom ? (
      <ImageTools>
        <img
          className={cs(s.downloadable, className)}
          src={objectURL}
          alt={objectURL}
          ref={ref}
          {...(download
            ? { onClick: this.handelDownload }
            : {
                onClick: () => {
                  this.props.onClick(objectURL);
                },
              })}
        />
      </ImageTools>
    ) : (
      <img
        className={cs(s.downloadable, className)}
        src={objectURL}
        alt={objectURL}
        ref={ref}
        {...(download
          ? { onClick: this.handelDownload }
          : {
              onClick: () => {
                this.props.onClick(objectURL);
              },
            })}
      />
    );
  };

  render() {
    const { objectToken, getTokenBefore } = this.props;
    if (!objectToken && !getTokenBefore) {
      return <span> نمایش این فایل پشتیبانی نمی‌شود!</span>;
    }

    return this.renderImg();
  }
}

LazyImage.propTypes = {
  className: PropTypes.string,
  path: PropTypes.string, // file address like: image.jpg
  objectToken: PropTypes.object,
  getDocAction: PropTypes.func.isRequired,
  getFileTokenAction: PropTypes.func.isRequired,
  onClick: PropTypes.func, // get file objectURL on image click
  download: PropTypes.bool, // if true: file will save on image click
  ref: PropTypes.func,
  hasZoom: PropTypes.bool,
  getTokenBefore: PropTypes.bool,
};

LazyImage.defaultProps = {
  className: '',
  objectToken: null,
  path: null,
  onClick: () => {},
  download: false,
  ref: () => {},
  hasZoom: false,
  getTokenBefore: false,
};

const mapState = () => ({});

const mapDispatch = {
  getDocAction,
  getFileTokenAction,
};

export default connect(mapState, mapDispatch)(withStyles(s)(LazyImage));
export const LazyImageTest = LazyImage;
