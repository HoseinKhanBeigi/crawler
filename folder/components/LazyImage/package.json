{
  "name": "LazyImage",
  "version": "1.0.0",
  "private": true,
  "main": "./LazyImage.js"
}
