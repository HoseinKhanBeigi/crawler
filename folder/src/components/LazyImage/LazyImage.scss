@import '../variables';

.downloadable {
  max-height: 500px;
}

.noData {
  height: 100%;
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
}
