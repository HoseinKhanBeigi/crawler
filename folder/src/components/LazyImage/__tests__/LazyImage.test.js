/* eslint-env jest */
/* eslint-disable padded-blocks, no-unused-expressions */

import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { LazyImageTest as LazyImage } from '../LazyImage';

configure({ adapter: new Adapter() });

describe('LazyImage Test', () => {
  let wrapper;
  beforeEach(() => {
    const props = {
      getDocAction: () => {},
    };
    wrapper = shallow(<LazyImage className="Test" {...props} />);
  });

  it('should render children when passed in', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
