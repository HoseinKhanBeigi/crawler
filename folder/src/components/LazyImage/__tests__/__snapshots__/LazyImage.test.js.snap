// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LazyImage Test should render children when passed in 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LazyImage
    className="Test"
    download={false}
    getDocAction={[Function]}
    objectToken={null}
    onClick={[Function]}
    path={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": "span",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": "span",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
